File: backend/services/geo_service.py
"""Geographic utility functions for distance and coordinate calculations.

Provides Haversine distance computation, radius checks, bounding-box
generation, and coordinate validation used by the discovery engine
and geo-verified check-ins.
"""
import math

def calculate_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Calculate the Haversine distance between two coordinates in kilometers.

    Args:
        lat1/lng1 (float): First point coordinates.
        lat2/lng2 (float): Second point coordinates.

    Returns:
        float: Distance in kilometers, rounded to 2 decimal places.
    """
    EARTH_RADIUS_KM = 6371.0
    lat1_rad = math.radians(lat1)
    lng1_rad = math.radians(lng1)
    lat2_rad = math.radians(lat2)
    lng2_rad = math.radians(lng2)
    dlat = lat2_rad - lat1_rad
    dlng = lng2_rad - lng1_rad
    a = math.sin(dlat / 2)**2 + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlng / 2)**2
    c = 2 * math.asin(math.sqrt(a))
    distance_km = EARTH_RADIUS_KM * c
    return round(distance_km, 2)

def is_within_radius(lat1: float, lng1: float, lat2: float, lng2: float, radius_km: float) -> bool:
    """Check whether two coordinates are within a given radius.

    Args:
        lat1/lng1 (float): First point coordinates.
        lat2/lng2 (float): Second point coordinates.
        radius_km (float): Maximum distance in kilometers.

    Returns:
        bool: True if the distance is within the radius.
    """
    distance = calculate_distance(lat1, lng1, lat2, lng2)
    return distance <= radius_km

def get_bounding_box(lat: float, lng: float, radius_km: float) -> dict:
    """Compute a rectangular bounding box around a point for geo queries.

    Args:
        lat (float): Center latitude.
        lng (float): Center longitude.
        radius_km (float): Radius in kilometers.

    Returns:
        dict: ``{"min_lat": float, "max_lat": float, "min_lng": float, "max_lng": float}``
    """
    EARTH_RADIUS_KM = 6371.0
    angular_distance = radius_km / EARTH_RADIUS_KM
    min_lat = lat - math.degrees(angular_distance)
    max_lat = lat + math.degrees(angular_distance)
    min_lng = lng - math.degrees(angular_distance / math.cos(math.radians(lat)))
    max_lng = lng + math.degrees(angular_distance / math.cos(math.radians(lat)))
    return {
        "min_lat": round(min_lat, 6),
        "max_lat": round(max_lat, 6),
        "min_lng": round(min_lng, 6),
        "max_lng": round(max_lng, 6)
    }

def validate_coordinates(lat: float, lng: float) -> bool:
    """Validate that latitude and longitude are within valid ranges.

    Args:
        lat (float): Latitude (-90 to 90).
        lng (float): Longitude (-180 to 180).

    Returns:
        bool: True if both coordinates are valid.
    """
    return -90 <= lat <= 90 and -180 <= lng <= 180

File: backend/src/services/geo_service.rs
/// Haversine distance in kilometres between two lat/lng points.
pub fn haversine_km(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    const R: f64 = 6371.0;
    let dlat = (lat2 - lat1).to_radians();
    let dlon = (lon2 - lon1).to_radians();
    let a = (dlat / 2.0).sin().powi(2)
        + lat1.to_radians().cos() * lat2.to_radians().cos() * (dlon / 2.0).sin().powi(2);
    2.0 * R * a.sqrt().asin()
}

/// Encode a lat/lng pair into a geo cell string at the given precision.
pub fn geo_cell(lat: f64, lng: f64, precision: u32) -> String {
    let factor = 10_f64.powi(precision as i32);
    let cell_lat = (lat * factor).floor() / factor;
    let cell_lng = (lng * factor).floor() / factor;
    format!("{:.prec$}:{:.prec$}", cell_lat, cell_lng, prec = precision as usize)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_haversine_same_point() {
        assert!(haversine_km(37.77, -122.42, 37.77, -122.42) < 0.001);
    }

    #[test]
    fn test_haversine_known_distance() {
        // SF to Oakland ≈ 13km
        let d = haversine_km(37.7749, -122.4194, 37.8044, -122.2712);
        assert!((d - 13.0).abs() < 1.5);
    }
}
